spec: update cli sign spec for tag to digest translation (#439)
notaryproject/notation: specs/commandline/sign.md

Use `notation sign` to sign artifacts.

> Signs an OCI artifact that is stored in a registry. Always use a `digest` to identify an artifact. `Tags` are mutable, but `digests` uniquely and immutably identify artifacts. If a tag is used, notation resolves the tag to the `digest` before signing.
> 
> Upon successful signing, the generated signature is pushed to the registry and associated with the signed OCI artifact. The output message is printed out as following:

Reviewer: ```suggestion
Signs an OCI artifact stored in the registry. Always sign artifact using digest(`@sha256:...`) rather than a tag(`:latest`) because tags are mutable and you might end up signing something that you didnt intend to. If a tag is used, notation resolves the tag to the digest before signing.

Upon successful signing, the generated signature will be pushed to the registry and output will be:
```

Author: - Update the "you might end up signing something that you didn't intend to" to "and a tag reference can point to a different artifact than the one signed ", which is more aligned with "verify CLI".

- I would like to keep the "present tense", not use "future tense".

```text

> Sign succeeded. Signature has been attached to <registry>/<repository>@<digest>.

Reviewer 2: This can be improved. I would recommend following some guidelines (here a good example https://spark.apache.org/error-message-guidelines.html). I would recommend the following:

```Signing of artifact <artifact-as-provided-in-the-command> succeeded and the signature is attached to <registry>/<repository>@<digest>.```

> Successfully signed <registry>/<repository>@<digest>.

Reviewer 3: It seems not consistent with `notation verify` that the sentence has a dot at the end or not.

> Signing of artifact <reference> succeeded and the signature is attached to <registry>/<repository>@<digest>.

Reviewer: reference is the redundant information.
```suggestion
Successfully signed <registry>/<repository>@<digest>
```

```

If a `tag` is used to identify the OCI artifact, the output message is as following:

```test

> Warning: Tag is used. Always use digest to identify the reference uniquely and immutably.

Reviewer 2: I think we can improve on the error message. I would recommend following some guidelines (here a good example https://spark.apache.org/error-message-guidelines.html). I would recommend improving this as follows:

```
Warning: A tag is used to identify the artifact for signing. Artifact tags are mutable. Use digests to uniquely identify artifacts and avoid mutability.
Resolving artifact tag <tag> to digest <digest> before signing.
Signing of artifact <artifact-as-provided-in-the-command> succeeded and the signature is attached to <registry>/<repository>@<digest>.
```

Also, this seems to be specific to the image signing. What about other artifacts?

Author: Thanks Toddy for sharing the guidelines. It is very useful. (Finally, I got the guidelines, I was searching for a while)

I have questions on `<artifact-as-provided-in-the-command>`, once tag is resolved to digest, should we use the same provided in the command or just replacing the tag with digest?

Author: Currently we only support signing remote artifacts, so the artifacts are identified by <registry>/<repository>@<digest>. It is the same signing logical as container images. Is there anything you see from output message indicates the artifact is a container image?

Reviewer 2: Regarding the first comment, the reason I typed `<artifact-as-provided-in-the-command>` is because the user can type the tag for the artifact in the command. Here are the steps:
```
notation sign test.mycr.io/myimage:tag
```
The message should look like:
```
Warning: A tag is used to identify the artifact for signing. Artifact tags are mutable. Use digests to uniquely identify artifacts and avoid mutability.
Resolving artifact tag 'tag' to digest 'sha:1234567890` before signing.
Signing of artifact test.mycr.io/myimage:tag succeeded and the signature is attached to test.mycr.io/myimage@sha1234567890.
```
This way, we give the user information what they typed and what changes were made as part of the command. Hope this helps.

> Resolved artifact tag '<tag>' to digest '<digest>' before signing.

Reviewer 3: Note: These outputs will be in the form of logs.

> Resolved artifact tag '<tag>' to digest '<digest>' before signing.

Reviewer 4: nit:
```suggestion
Resolved artifact tag `<tag>` to digest `<digest>` before signing.
```

> Resolved artifact tag '<tag>' to digest '<digest>' before signing.

Reviewer 4: Also, I noticed that, for Sign, we are printing out the warning before the resolved message. Whereas in Verify, we are printing out the warning after the resolved message. Is this by design?

Author: Yes.

> Warning: A tag is used to identify the artifact for signing. Artifact tags are mutable. Use digests to uniquely identify artifacts and avoid mutability.
> Resolving artifact tag '<tag>' to digest '<digest>' before signing.
> Signing of artifact <reference> succeeded and the signature is attached to <registry>/<repository>@<digest>.

Reviewer: ```suggestion
Warning: Always sign artifact using digest(`@sha256:...`) rather than a tag(`:latest`) because tags are mutable and you might end up signing something that you didnt intend to.
Successfully signed <registry>/<repository>@<digest>
```

Author: Suggest using the following warning message:

"Warning: Always sign the artifact using digest(`@sha256:...`) rather than a tag(`:latest`) because tags are mutable and a tag reference can point to a different artifact than the one signed."

## Usage

### Sign an OCI artifact

```shell
# Prerequisites: 

> # - A signing plugin is installed. See plugin documentation (https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md) for more details.

Reviewer 3: Instead of the main branch, should we use a tag version for the URL?

Author: Can we update it later once we have cut a version for notaryproject?

## Usage

### Sign an OCI artifact

```shell
# Prerequisites: 

> # - A signing plugin is installed. See plugin documentation (https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md) for more details.

Reviewer 5: I have two concerns:
1. A signing plugin is not a hard prerequisite for all cases, only signing with the key stored in the KMS system.
2. This link is not a good reference for installing a plugin.

Author: - We don't support local signing scenario in rc.1
- We can replace it with an installation guide, but we don't have it yet. The link here is for user to learn what is a plugin

# - Configure the signing plugin as instructed by plugin vendor.

# Add a default signing key referencing the remote key identifier, and the plugin associated with it.
notation key add --default --name <key_name> --plugin <plugin_name> --id <remote_key_id>

# sign an artifact stored in a registry
notation sign <registry>/<repository>@<digest>
```

An example for a successful signing:

> ```shell

Reviewer 3: ```suggestion
```console
```

$ notation sign localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9
Successfully signed localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9.
```

### Sign an OCI artifact using COSE signature format

```shell
# Prerequisites: 
# A default signing key is configured using CLI "notation key"

# Use option "--signature-format" to set the signature format to COSE.
$ notation sign --signature-format cose <registry>/<repository>@<digest>
```

### Sign an OCI artifact stored in a registry using the default signing key

```shell
# Prerequisites: 
# A default signing key is configured using CLI "notation key"

# Use a digest that uniquely and immutably identifies an OCI artifact.

> $ notation sign <registry>/<repository>@<digest>

Reviewer 3: There is no `$` in shell.

```

### Sign an OCI artifact stored in a registry and specify the signature expiry duration, for example 24 hours

```shell
notation sign --expiry 24h <registry>/<repository>@<digest>
```

### Sign an OCI artifact stored in a registry using a specified signing key

```shell
# List signing keys to get the key name

> $ notation key list

Reviewer 3: same here and the lines below

# Sign a container image using the specified key name
$ notation sign --key <key_name> <registry>/<repository>@<digest>
```

### Sign an OCI artifact identified by a tag

```shell
# Prerequisites: 
# A default signing key is configured using CLI "notation key"

# Use a tag to identify a container image
$ notation sign <registry>/<repository>:<tag>
```

An example for a successful signing:

```shell
$ notation sign localhost:5000/net-monitor:v1

> Warning: Always sign the artifact using digest(`@sha256:...`) rather than a tag(`:latest`) because tags are mutable and a tag reference can point to a different artifact than the one signed.

Reviewer 3: ```suggestion
Warning: Always sign the artifact using digest(`@sha256:...`) rather than a tag(`:v1`) because tags are mutable and a tag reference can point to a different artifact than the one signed.
```

> Resolving artifact tag 'v1' to digest 'sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9' before signing.

Reviewer: Do we need to add this line? since next line already shows the digest.

Author: I think so, this is to indicate clearly that it is the supplied tag resolved to a digest.

## Usage

> ### Sign an OCI artifact stored in a registry using a remote key

Reviewer: As of RC1 we only support remote key so we dont need to call that out.
```suggestion
### Sign an OCI artifact
```

```shell
# Prerequisites: 

> # - A compliant signing plugin is installed in notation. See notation plugin documentation (https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md) for more details.

Reviewer: ```suggestion
# - A signing plugin is installed, dee [plugin documentation](https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md) for more details.
```

Author: It's in the comment, so the `[]()` format doesn't work.

> # - User creates keys and certificates in a 3rd party key provider (e.g. key vault, key management service). The signing plugin installed in previous step must support generating signatures using this key provider.

Reviewer: ```suggestion
# - Configure the signing plugin as instructed by plugin vender
```

> # Add a default signing key referencing the key identifier for the remote key, and the plugin associated with it.

Reviewer: ```suggestion
# Add a default signing key referencing the remote key identifier and the plugin associated with it.
```

notation key add --default --name <key_name> --plugin <plugin_name> --id <remote_key_id>

> # sign an artifact stored in a registry using a remote key

Reviewer: ```suggestion
# sign an artifact stored in a registry
```

notation sign <registry>/<repository>@<digest>
```

An example for a successful signing:

```shell
$ notation sign localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9
Signing of artifact <reference> succeeded and the signature is attached to localhost:5000/net-monitor@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9.
```

### Sign an OCI artifact using COSE signature format

```shell
# Prerequisites: 
# A default signing key is configured using CLI "notation key"

# Use option "--signature-format" to set the signature format to COSE.
$ notation sign --signature-format cose <registry>/<repository>@<digest>
```

### Sign an OCI artifact stored in a registry using the default signing key

```shell
# Prerequisites: 
# A default signing key is configured using CLI "notation key"

# Use a digest that uniquely and immutably identifies an OCI artifact.
$ notation sign <registry>/<repository>@<digest>
```

### Sign an OCI artifact stored in a registry and specify the signature expiry duration, for example 24 hours

```shell

> notation sign --expiry 24h <registry>/<repository>@<digest>

Reviewer: IMO day makes more sense because user will rarely want an signature to expire in less than a day.

Author: Currently `day` is not support yet.

Reviewer: Opened issue https://github.com/notaryproject/notation/issues/451 and tagged it for rc2

Author: Thanks Pritesh. There was an issue https://github.com/notaryproject/notation/issues/423 I created earlier.

```

### Sign an OCI artifact stored in a registry using a specified signing key

```shell
# List signing keys to get the key name
$ notation key list

# Sign a container image using the specified key name
$ notation sign --key <key_name> <registry>/<repository>@<digest>
```

> ### Sign an OCI artifact identified by a tag

Reviewer: Will move this up with `Sign an OCI artifact `

Author: Since `tag` is not recommended way to identify the artifact, so it's better not to put it as a first usage.

notation sign <registry>/<repository>@<digest>
```

### Sign an OCI artifact stored in a registry and specify the signature expiry duration, for example 24 hours

```shell
notation sign --expiry 24h <registry>/<repository>@<digest>
```

### Sign an OCI artifact stored in a registry using a specified signing key

```shell
# List signing keys to get the key name
notation key list

# Sign a container image using the specified key name
notation sign --key <key_name> <registry>/<repository>@<digest>
```

### Sign an OCI artifact identified by a tag

```shell
# Prerequisites: 
# A default signing key is configured using CLI "notation key"

# Use a tag to identify a container image
notation sign <registry>/<repository>:<tag>
```

An example for a successful signing:

```console
$ notation sign localhost:5000/net-monitor:v1

> Warning: Always sign the artifact using digest(`@sha256:...`) rather than a tag(`:v1`) because tags are mutable and a tag reference can point to a different artifact than the one signed.

Reviewer 3: It should be the tag input from caller

> Warning: Always sign the artifact using digest(`@sha256:...`) rather than a tag(`:v1`) because tags are mutable and a tag reference can point to a different artifact than the one signed.

Reviewer 3: ```suggestion
Warning: Always sign the artifact using digest(`@sha256:...`) rather than a tag(`:<tag>`) because tags are mutable and a tag reference can point to a different artifact than the one signed.
```

Author: I would keep line 115 as it is, but change the text in line 18 from `:v1` to `:<tag>`. What do you think?

> Resolved artifact tag 'v1' to digest 'sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9' before signing.

Reviewer 4: nit:
```suggestion
Resolved artifact tag `v1` to digest `sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9` before signing.
```
